### PR TITLE
frontend_pytorch_v0.0.3 mnist ops supported by pytorch backend

### PR DESCRIPTION
**Pull request by ethancaballero**
- ops from pytorch mnist example supported by pytorch backend
- test_pytorch_ops can support ops that return multiple outputs

### CODE REVIEW

File: myia/abstract/prim.py [hunk 1]
```diff
@@ -619,6 +655,54 @@ def prod(iterable):
...
+
+@standard_prim(P.scatter_add)
+async def _inf_scatter_add(self, engine, input, dim, index, src):
+    return input
+
+
+@standard_prim(P.argmax)
+async def _inf_argmax(self, engine, input, dim):
+    shp = ()
+    if dim.values[VALUE] is None:
```
[the remark this answers is not in the document]
Comment (breuleux) [reply myia/abstract/prim.py]: Also, argmax should not take None as the dim.

File: myia/frontends/pytorch.py [hunk 2]
```diff
@@ -102,12 +131,16 @@ def mod_sub(self, x):
...
 # Curently blacklists all constructors except '_parameters' and '_modules'.
 # 'training' should probably be removed from blacklist in next PR.
+"""
 blacklist = ('_backend', '_buffers', '_backward_hooks', '_forward_hooks',
              '_forward_pre_hooks', '_state_dict_hooks',
              '_load_state_dict_pre_hooks',
 
              'training'
              )
+             """
```
Comment (breuleux) [inline myia/frontends/pytorch.py]: Remove comment.

File: myia/frontends/pytorch.py [hunk 3]
```diff
@@ -19,60 +20,88 @@
...
 })
 
 
 standard_method_map[PyTorchTensor] = \
     standard_method_map[NDArray].copy()
 standard_method_map[PyTorchTensor].update({
     'dim': C.ndim,
-    'exp': C.array_exp,
+    'argmax': argmax,
+    'exp': C.exp,
```
Comment (breuleux) [inline myia/frontends/pytorch.py]: `C.array_exp` is the right one to use. Same comment for log and tanh below.

File: myia/frontends/pytorch.py [hunk 4]
```diff
@@ -117,17 +150,19 @@ def _to_abstract(self, v: torch.nn.Module, **kwargs):
...
-            # TODO: Remove "(isinstance(v, torch.nn.Sequential) and"
-            #       once Alias PR ready
-            # TODO: Remove rest of if statement once Dict supports empty Dict
-            if var_k not in ('_parameters', '_modules') or \
-                    (isinstance(v, torch.nn.Sequential) and
-                     var_v != OrderedDict()):
-                attrs[var_k] = self(var_v, **kwargs)
+    fields = {}
+    for var_k in dir(v):
+        if (var_k not in blacklist) or (var_k in ('_parameters', '_modules')):
```
Comment (breuleux) [inline myia/frontends/pytorch.py]: It would be better to simply remove `_parameters` and `_modules` from the blacklist when it is created.

File: myia/frontends/pytorch.py [hunk 5]
```diff
@@ -102,12 +131,16 @@ def mod_sub(self, x):
...
+"""
 blacklist = ('_backend', '_buffers', '_backward_hooks', '_forward_hooks',
              '_forward_pre_hooks', '_state_dict_hooks',
              '_load_state_dict_pre_hooks',
 
              'training'
              )
+             """
+blacklist = set(dir(torch.nn.Module()) + ["__constants__"])
+blacklist.add('reset_parameters')
```
Comment (breuleux) [inline myia/frontends/pytorch.py]: You're adding two entries to the blacklist in two different ways (+ and .add), which is not very consistent.

File: myia/pipeline/resources.py [hunk 6]
```diff
@@ -227,8 +227,8 @@
         '__ge__': C.array_ge,
         '__matmul__': P.dot,
         '__len__': M.array_len,
-        '__getitem__': P.array_getitem,
-        '__setitem__': P.array_setitem,
+        '__getitem__': C.array_getitem_wrap,
+        # '__setitem__': C.array_setitem_wrap,
```
Comment (breuleux) [inline myia/pipeline/resources.py]: We don't actually ever want to associate anything to `__setitem__`.

```suggestion
```

File: myia/prim/grad_implementations.py [hunk 7]
```diff
@@ -16,15 +16,19 @@
 from ..pipeline import standard_pipeline
 from ..utils import InternalInferenceError, Registry, newenv
 from . import ops as primops
-from .py_implementations import (
+from .py_implementations import (  # array_cast,
```
Comment (breuleux) [inline myia/prim/grad_implementations.py]:
```suggestion
from .py_implementations import (
```

File: tests/frontends/test_pytorch.py [hunk 8]
```diff
@@ -983,6 +988,127 @@ def step(model, inp):
...
+    if bias.grad is not None:
+        bias.grad.data.zero_()
+
+    pt_cost.backward()
+    #"""
+
+    #assert torch.allclose(my_out_dinp_grad, input.grad.data)
+    #assert torch.allclose(my_out_dw_grad, weight.grad.data)
+    #assert torch.allclose(my_out_db_grad, bias.grad.data)
+#'''
```
Comment (breuleux) [inline tests/frontends/test_pytorch.py]: Remove comment.

Comment (ethancaballero) [reply tests/frontends/test_pytorch.py]: This is going to be used in next PR when Heterogeneous broaden.

File: myia/frontends/pytorch.py (hunk 4 above)
Comment (ethancaballero) [reply myia/frontends/pytorch.py]: This portion of if statement will be removed when alias for classes/module is supported in next PR. It's easier for me remember to remove it when it looks awkward in if statement (like it currently does).

File: tests/frontends/test_pytorch.py (hunk 8 above)
Comment (breuleux) [reply tests/frontends/test_pytorch.py]: I would prefer it if you stashed it or committed it on a different branch, so that the master version remains clean.

File: myia/frontends/pytorch.py (hunk 4 above)
Comment (breuleux) [reply myia/frontends/pytorch.py]: Okay, add a comment to explain this.
